File: leetcode-js/152-maximum-product-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    'use strict';

    let result;
    let max;
    let min;

    result = max = min = nums[0];

    for (let i = 1; i < nums.length; i += 1) {
        const curMax = max;
        max = Math.max(curMax * nums[i], min * nums[i], nums[i]);
        min = Math.min(curMax * nums[i], min * nums[i], nums[i]);
        result = Math.max(result, max);
    }

    return result;
};

File: leetcode-java/152-maximum-product-subarray.java
public class Solution {
    public int maxProduct(int[] nums) {
        int max = nums[0];
        int min = nums[0];
        int res = nums[0];

        for (int i = 1; i < nums.length; i++) {
            int n = nums[i];
            int curMax = max;

            max = Math.max(Math.max(n, curMax * n), min * n);
            min = Math.min(Math.min(n, curMax * n), min * n);

            res = Math.max(res, max);
        }

        return res;
    }
}
